rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,752 @@
+    """
+    # Operation
+    # ---------
+    # While the merge is still better than `min_goodness` we determine which
+    # entries below it in the table it covers. For each of these covered
+    # entries we find which bits are Xs in the merged entry and are NOT Xs in
+    # the covered entry.
+    #
+    # For example:
+    #
+    #     Merged entry:      ...0XXX1...
+    #     Covered entry:     ...010XX...
+    #     Bits of interest:      ^^
+    #     Label used below:      mn
+    #
+    # In this case there are 2 ``settable`` bits of interest highlighted.
+    # Whenever we have multiple covered entries we care only about the entries
+    # with the fewest number of ``settable`` bits because these most constrain
+    # which entries we may remove from the merge to avoid covering up the lower
+    # entry.
+    #
+    # NOTE:
+    #   * If there is only 1 ``settable`` bit then we are very constrained in
+    #     terms of which entries must be removed from the merge to avoid
+    #     covering a lower entry.
+    #   * If there are no ``settable`` bits then we cannot possibly avoid
+    #     covering the lower entry - the only correct action is to return an
+    #     empty merge.
+    #
+    # Assuming that there were no covered entries with no ``settable`` bits
+    # (that is ``stringency > 0``) then ``bits_and_vals`` contains pairs of
+    # bits and boolean values which indicate which values need to be removed
+    # from which bit positions to avoid covering up lower entries. If the
+    # example above were the only covered entry then ``bits_and_vals`` would
+    # contain ``(m, True)`` to indicate that all entries containing Xs or 1s in
+    # the left-most bit of interest could be removed to avoid the covered entry
+    # and ``(n, False)`` to indicate that all entries containing Xs or 0s in
+    # the right-most bit of interest could be removed to avoid covering the
+    # entry.
+    #
+    # NOTE:
+    #   ``bits_and_vals`` consists of a set of options (e.g., we *could* remove
+    #   all entries with Xs or 1s in bit ``m`` *or* we could remove all entries
+    #   with Xs or 0s in bit ``n``, either would resolve the above covering).
+    #
+    # To determine which course of action to take we build a dictionary mapping
+    # each of the pairs in ``bits_and_vals`` to the entries that would need to
+    # be removed to "set" that bit in the merged entry. For example, we might
+    # end up with:
+    #
+    #     options = {(m, True): {1, 4, 5},
+    #                (n, False): {3, 7}}
+    #
+    # Indicating that we'd need to remove entries 1, 4 and 5 from the merge to
+    # "set" the mth bit of the merged to 0 or that we'd need to remove entries
+    # 3 and 7 to set the nth bit of the merged entry to set the nth bit to 1.
+    #
+    # NOTE:
+    #   The boolean part of the pair indicates which value needs to be removed
+    #   (True -> remove all 1s; False -> remove all 0s). If all Xs and 1s in a
```

Comment:
"remove all 1s **and Xs**" and "remove all 0s **and Xs**"? Might be clearer than stating it separately later.